src/js_subscriptions.hpp

```diff
@@ -0,0 +1,885 @@
+    auto query = results->get_query();
+
+    auto it = subs->find(query);
+
+    if (it == subs->end()) {
+        return_value.set_null();
+    }
+    else {
+        auto sub = SubscriptionClass<T>::create_instance(ctx, *it);
+        return_value.set(sub);
+    }
+}
+
+/**
+ * @brief See `wait_for_synchronization_impl`
+ *
+ * @param ctx JS context
+ * @param this_object \ref ObjectType wrapping the SubscriptionSet
+ * @param args \ref A single argument containing a callback to be called when the state is "Complete" or "Error"
+ * @param return_value \ref None
+ */
+template <typename T>
+void SubscriptionSetClass<T>::wait_for_synchronization(ContextType ctx, ObjectType this_object, Arguments& args,
+                                                       ReturnValue& return_value)
+{
+    args.validate_count(1);
+    auto callback = Value::validated_to_function(ctx, args[0], "callback");
+
+    Protected<FunctionType> protected_callback(ctx, callback);
+    Protected<ObjectType> protected_this(ctx, this_object);
+    Protected<typename T::GlobalContext> protected_ctx(Context<T>::get_global_context(ctx));
+
+    SubscriptionSetClass<T>::wait_for_synchronization_impl(protected_ctx, protected_this, protected_callback);
+}
+
+/**
+ * @brief Invoke a callback when the SubscriptionSet's state becomes "Complete" or "Error".
+ * Will invoke it immediately if the state is already "Complete". Will return an error to
+ * the callback if the state is or becomes "Error", or if it is called on before creating
+ * any subscriptions.
+ *
+ * @param ctx JS context
+ * @param this_object \ref ObjectType wrapping the SubscriptionSet
+ * @param callback \ref Callback function to be invoked when the state is "Complete" or "Error"
+ */
+template <typename T>
+void SubscriptionSetClass<T>::wait_for_synchronization_impl(ContextType ctx, ObjectType this_object,
+                                                            FunctionType callback)
+{
+    Protected<FunctionType> protected_callback(ctx, callback);
+    Protected<ObjectType> protected_this(ctx, this_object);
+    Protected<typename T::GlobalContext> protected_ctx(js::Context<T>::get_global_context(ctx));
+
+    auto subs = get_internal<T, SubscriptionSetClass<T>>(protected_ctx, protected_this);
+
+    // Hold a weak_ptr to the SyncSession, so we can check if it still exists when our callback fires
+    // – if the Realm has gone out of scope and been garbage collected by the time the callback fires
+    // (which happens in tests), we get a crash otherwise
+    std::weak_ptr<SyncSession> sync_session = subs->sync_session;
+
+    std::function<StateChangeHandler> state_change_func;
+
+    util::EventLoopDispatcher<StateChangeHandler> state_change_handler(
+        [=](StatusWith<realm::sync::SubscriptionSet::State> state) {
+            HANDLESCOPE(protected_ctx)
+
+            // If the SyncSession has already closed, don't do anything as we will crash
+            if (sync_session.lock()) {
```

Comment:
I was wondering if it would make sense to reject the promise if this check fails?

Reply:
Yeah I guess it might do. The specific situation I was working around was in the tests, if you do: `subs.update(..)` without awaiting the result, then the test has finished by the time the `wait_for_synchronisation` that `subs.update` returns reaches this point, because `subs` is now out of scope (the test has finished executing) – so the return value of the promise was irrelevant. 

But I think it would make sense to reject it regardless, in case of some other situation that can cause this.